golem-shard-manager/src/shard_management.rs

```diff
@@ -129,63 +124,67 @@ impl ShardManagement {
+            let mut rebalance = {
+                let mut current_routing_table = routing_table.write().await;
 
-                    info!("Registered worker executor returned: {pod}")
-                } else {
-                    // New pod, adding with empty assignment
-                    current_routing_table.add_pod(&pod);
+                for pod in removed_pods {
+                    current_routing_table.remove_pod(&pod);
+                    info!(pod= %pod, "Pod removed");
+                }
 
-                    info!("Registered new worker executor: {pod}")
+                let mut send_full_assignment = Vec::new();
+                for pod in new_pods {
+                    if current_routing_table.has_pod(&pod) {
+                        // This pod has already an assignment - we have to send the full list of assigned shards to it
+                        send_full_assignment.push(pod.clone());
+                        info!(pod= %pod, "Pod returned");
+                    } else {
+                        // New pod, adding with empty assignment
+                        current_routing_table.add_pod(&pod);
+                        info!(pod= %pod, "Pod added");
+                    }
                 }
-            }
-            let mut rebalance = Rebalance::from_routing_table(&current_routing_table, threshold);
+                let mut rebalance =
+                    Rebalance::from_routing_table(&current_routing_table, threshold);
 
-            for pod in send_full_assignment {
-                let assignments = current_routing_table.get_shards(&pod).unwrap_or_default();
-                rebalance.add_assignments(&pod, assignments);
-            }
+                for pod in send_full_assignment {
+                    let assignments = current_routing_table.get_shards(&pod).unwrap_or_default();
+                    rebalance.add_assignments(&pod, assignments);
+                }
 
-            debug!("Applying rebalance plan: {rebalance}");
+                persistence_service
+                    .write(&current_routing_table)
+                    .await
+                    .expect("Failed to persist routing table after pod changes");
 
-            // Panicking in case any of the rebalancing steps fail (after some internal retries within those steps).
-            // This causes the shard manager to get restarted and have and retry the rebalance on next startup.
+                rebalance
+            };
 
-            persistence_service
-                .write(&current_routing_table, &rebalance)
-                .await
-                .expect("Failed to persist routing table");
-            drop(current_routing_table); // not holding the write lock while executing the rebalance
-
-            Self::execute_rebalance(worker_executors.clone(), &mut rebalance)
-                .await
-                .expect("Failed to execute rebalance");
+            debug!(rebalance=?rebalance, "Applying rebalance plan");
```

Comment:
here should be possible to use Display instead of Debug for arguments

Reply:
https://github.com/golemcloud/golem/pull/690/commits/0a86347c73d43333b0b3c22abe0447b3fe41bebe